src/diagnostics.rs

```diff
@@ -73,12 +78,42 @@ impl DiagnosticData {
             os_version,
             triple: target_lexicon::HOST.to_string(),
             is_ci,
-            failure_variant: None,
+            failure_chain: None,
         }
     }
 
-    pub fn variant(mut self, variant: String) -> Self {
-        self.failure_variant = Some(variant);
+    pub fn failure(mut self, err: &NixInstallerError) -> Self {
+        let mut failure_chain = vec![];
+        let diagnostic = err.diagnostic();
+        failure_chain.push(diagnostic);
+
+        let mut walker: &dyn std::error::Error = &err;
+        while let Some(source) = walker.source() {
+            if let Some(downcasted) = source.downcast_ref::<ActionError>() {
```

Comment:
Will we need to add a new branch for every new `Error` enum we add? I wonder if it would be possible to rely on `ActionError::Child` so we wouldn't have to remember...

If not, is there any way we could rely on the compiler to make sure we've attempted to downcast all possible error sources?

Reply:
> Will we need to add a new branch for every new Error enum we add? 

Only ones we care about for diagnostics!

> I wonder if it would be possible to rely on ActionError::Child so we wouldn't have to remember...

I don't really understand what you mean here! Can you help?

> If not, is there any way we could rely on the compiler to make sure we've attempted to downcast all possible error sources?

Hmmm not really since this could be quite literally any error due to how this API works, including ones we don't care about (like `std::io::Error`)

Reply:
Yeah, I'm not really sure what I was going for here... lol.

Some way to lean on the compiler to ensure we didn't leave anything out would have been great, but seeing as how we might not want diagnostics for every error as you say, this is probably not doable / useful.